### PR TITLE
Allow to validate a custom schema with `webpack.validate`

### PR DESCRIPTION
**Pull request by SpaceK33z**
**Please check if the PR fulfills these requirements**
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Other... Please describe:

**What is the current behavior?** (You can also link to an open issue here)

With `webpack.validate` it is possible to validate webpack options.

Example: `webpack.validate({ entry: {} })`

**What is the new behavior?**

It is now possible to add a second parameter to provide a custom schema. This is necessary for webpack-dev-server and webpack-dev-middleware, so they can provide their own schema without re-implementing the whole validation logic.

Usage in webpack-dev-* would be like this:

```js
webpack.validate({ host: 'localhost', port: 8000 }, webpackDevServerSchema);
```

**Does this PR introduce a breaking change?**
- [ ] Yes
- [x] No

**Other information**:

I would still need to make the actual schema for webpack-dev-*, but I wanted to do this first so I know how to code the rest.

### CODE REVIEW

File: lib/validateWebpackOptions.js [hunk 1]
```diff
@@ -9,11 +9,11 @@ var ajv = new Ajv({
...
 	verbose: true
 });
-var validate = ajv.compile(webpackOptionsSchema);
 
-function validateWebpackOptions(options) {
+function validateWebpackOptions(options, customSchema) {
+	var schema = customSchema || webpackOptionsSchema;
 	if(Array.isArray(options)) {
-		var errors = options.map(validateObject);
+		var errors = options.map(validateObject.bind(this, schema));
```
Comment (sokra) [inline lib/validateWebpackOptions.js]: I think you need to switch the argument order for this to work.

File: lib/validateWebpackOptions.js [hunk 2]
```diff
@@ -9,11 +9,11 @@ var ajv = new Ajv({
 	allErrors: true,
 	verbose: true
 });
-var validate = ajv.compile(webpackOptionsSchema);
 
-function validateWebpackOptions(options) {
+function validateWebpackOptions(options, customSchema) {
+	var schema = customSchema || webpackOptionsSchema;
```
Comment (sokra) [inline lib/validateWebpackOptions.js]: better remove the default and always pass the schema. remove the require. pass the schema from caller side. rename the function and file to `validateOptions`.

**Comment by SpaceK33z** [comment]
@sokra, I implemented your feedback + the tests succeed again :).

**Comment by sokra** [comment]
Could you rename the file to `validateSchema` and expose it as `webpack.validateSchema(schema, obj)`. And add a `webpack.validate(obj)` which behaves as before (validating with the webpack schema).

**Comment by SpaceK33z** [comment]
Done. What do you think of the implementation for WDS: https://github.com/webpack/webpack-dev-server/commit/e9c4ceb9bfb91b01c61d0167867e921dd0b0b98f? I have to access `webpack/lib/WebpackOptionsValidationError` from there, which is a bit nasty. I could expose it on the `webpack` export. However, `WebpackOptionsValidationError.js` contains some webpack-specific error messages like "The 'debug' property was removed in webpack 2.", so maybe I should make my own error formatter.

**Comment by SpaceK33z** [comment]
I've made a `WebpackOptionsValidationError` variant for WDS now, so you can ignore the last part of my previous comment.

**Comment by SpaceK33z** [comment]
@sokra, implemented your feedback from Slack. After this, WDS config validation works :).

File: lib/validationSchema.js [hunk 3]
```diff
@@ -2,18 +2,16 @@
...
 var Ajv = require("ajv");
 var ajv = new Ajv({
 	errorDataPath: "configuration",
 	allErrors: true,
 	verbose: true
 });
-var validate = ajv.compile(webpackOptionsSchema);
 
-function validateWebpackOptions(options) {
+function validateWebpackOptions(schema, options) {
```
Comment (sokra) [inline lib/validationSchema.js]: Could you rename this function to `validateSchema` too?

**Comment by SpaceK33z** [comment]
@sokra, sorry, should've seen that. It's fixed now.

File: lib/validationSchema.js [hunk 4]
```diff
@@ -2,18 +2,16 @@
...
 var Ajv = require("ajv");
 var ajv = new Ajv({
 	errorDataPath: "configuration",
 	allErrors: true,
 	verbose: true
 });
-var validate = ajv.compile(webpackOptionsSchema);
 
-function validateWebpackOptions(options) {
+function validationSchema(schema, options) {
```
Comment (sokra) [inline lib/validationSchema.js]: validationSchema -> validateSchema

Comment (SpaceK33z) [inline lib/validationSchema.js]: Ah, I wanted to use the same name as the file, but fair enough. Fixed.

**Comment by SpaceK33z** [comment]
Implemented the function + file rename now (missed that you also wanted me to rename the file at first).

**Comment by sokra** [comment]
Thanks

**Pull request merged**